src/wp-includes/user.php

```diff
@@ -3690,42 +3738,73 @@ function send_confirmation_on_profile_email() {
-		 *     @type string $newemail The proposed new email address.
-		 * }
-		 */
-		$content = apply_filters( 'new_user_email_content', $email_text, $new_user_email );
+	/**
+	 * Filters the text of the email sent when a change of user email address is attempted.
+	 *
+	 * The following strings have a special meaning and will get replaced dynamically:
+	 * - ###USERNAME###  The current user's username.
+	 * - ###ADMIN_URL### The link to click on to confirm the email change.
+	 * - ###EMAIL###     The new email.
+	 * - ###SITENAME###  The name of the site.
+	 * - ###SITEURL###   The URL to the site.
+	 *
+	 * @since MU (3.0.0)
+	 * @since 4.9.0 This filter is no longer Multisite specific.
+	 *
+	 * @param string $email_text     Text in the email.
+	 * @param array  $new_user_email {
+	 *     Data relating to the new user email address.
+	 *
+	 *     @type string $hash     The secure hash used in the confirmation link URL.
+	 *     @type string $newemail The proposed new email address.
+	 * }
+	 */
+	$content = apply_filters( 'new_user_email_content', $email_text, $new_user_email );
 
-		$content = str_replace( '###USERNAME###', $current_user->user_login, $content );
-		$content = str_replace( '###ADMIN_URL###', esc_url( admin_url( 'profile.php?newuseremail=' . $hash ) ), $content );
-		$content = str_replace( '###EMAIL###', $_POST['email'], $content );
-		$content = str_replace( '###SITENAME###', $sitename, $content );
-		$content = str_replace( '###SITEURL###', home_url(), $content );
+	$content = str_replace( '###USERNAME###', $user->user_login, $content );
+	$content = str_replace( '###ADMIN_URL###', esc_url( $confirm_url ), $content );
+	$content = str_replace( '###EMAIL###', $email, $content );
+	$content = str_replace( '###SITENAME###', $sitename, $content );
+	$content = str_replace( '###SITEURL###', home_url(), $content );
 
-		/* translators: New email address notification email subject. %s: Site title. */
-		wp_mail( $_POST['email'], sprintf( __( '[%s] Email Change Request' ), $sitename ), $content );
+	/* translators: New email address notification email subject. %s: Site title. */
+	wp_mail( $email, sprintf( __( '[%s] Email Change Request' ), $sitename ), $content );
+
+	return true;
+}
 
-		$_POST['email'] = $current_user->user_email;
+/**
+ * Process the confirmation of an email change request.
+ *
+ * @since x.x
+ *
+ * @param int    $user_id    The User ID being acted upon.
+ * @param string $email_hash The email hash of the request
+ *
+ * @return bool Whether or not the change succeeded.
+ */
+function send_user_email_change_confirmation_process( $user_id, $email_hash ) {
+
+	$new_email = get_user_meta( $user_id, '_new_email', true );
+	if ( ! $new_email || ! hash_equals( $new_email['hash'], $email_hash ) ) {
+		return false;
+	}
+
+	$the_user         = get_user_by( 'id', $user_id );
+	$user             = new stdClass();
+	$user->ID         = $the_user->ID;
+	$user->user_email = esc_html( trim( $new_email['newemail'] ) );
```

Comment:
I'm not sure why this is `esc_html()`, this seems somewhat in error to me, but it's what the existing code does.